tests|farmer: Attempt to make `test_farmer_harvester_rpc.py` more reliable

xdustinface opened the pull request:
This fails a lot on CI but i am not exactly sure yet whats going on that it fails. One guess is that the cache update happens before the harvester loaded its plots so that the next update won't happen in time so this PR forces an cache update and triggers extra harvester plot refreshing.

xdustinface committed:
tests: Increase interval between harvester checks
To avoid spamming logs by excessive plot refreshing and cache updates.

xdustinface committed:
tests: Log error messages if `get_harvesters` returns unexpected data

xdustinface committed:
tests: Trigger plot refresh on the harvester in `test_get_harvesters`

arvidn commented on tests/core/test_farmer_harvester_rpc.py:
```diff
@@ -171,14 +171,18 @@ async def have_plots():
             # Test farmer get_harvesters
             async def test_get_harvesters():
+                harvester.plot_manager.trigger_refresh()
+                await farmer_api.farmer.update_cached_harvesters(force_update=True)
```
you don't think there's a risk that this might hide a real issue in the harvester?
I think it would be good to understand why the harvester's cache become stale. This won't happen in a real-world running of the harvester, right? and it needs to be able to pick up the plots anyway.

If you do understand why this fails and it would only ever happen in the test, I think there should be a comment here explaining it for future readers of this code.

xdustinface replied:
I don't think there is a risk. This just triggers the cache update earlier which happens usually with a 90s interval. I changed the code though to not have the force flag but instead just change the update interval.

xdustinface committed:
farmer|tests: Update cache more frequent for `test_get_harvesters`

xdustinface committed:
tests: Don't assert in `test_refresh_callback`

This lead to a crash of the plot refreshing thread and with this it 
silently failed. With this change its at least possible to see which 
value is incorrect.

arvidn commented on chia/farmer/farmer.py:
```diff
@@ -608,7 +612,7 @@ async def update_cached_harvesters(self) -> bool:
                 cache_entry.bump_last_update()
                 response = await connection.request_plots(
-                    harvester_protocol.RequestPlots(), timeout=UPDATE_HARVESTER_CACHE_INTERVAL
+                    harvester_protocol.RequestPlots(), timeout=self.update_harvester_cache_interval
```
it seems a bit odd to use the *cache* interval as the timeout to this request. It's not obvious that it makes sense.

xdustinface replied:
It's coming from https://github.com/Chia-Network/chia-blockchain/pull/7244.

arvidn commented on chia/farmer/farmer.py:
```diff
@@ -637,7 +641,7 @@ async def get_cached_harvesters(self, connection: WSChiaConnection) -> Harvester
         node_cache = host_cache.get(connection.peer_node_id.hex())
         if node_cache is None:
-            node_cache = HarvesterCacheEntry()
+            node_cache = HarvesterCacheEntry(self.update_harvester_cache_interval)
```
this seems to be the main use of this member. You don't think it would be simpler and more direct to just pass this value as a parameter to this function? (i.e. to `get_cached_harvesters()`)

arvidn commented on chia/farmer/farmer.py:
```diff
@@ -67,9 +67,10 @@
+    def __init__(self, update_interval: int):
         self.data: Optional[dict] = None
         self.last_update: float = 0
+        self.update_interval = update_interval
```
it seems a bit odd to store this interval in *every* cache entry. It's going to be the same value in all of them, right?

You could just pass it into `needs_update()` and `expired()` instead.

Alternatively, you could pass it into `bump_last_update()` and change the `last_update` timestamp to instead be *next_update* timestamp.

xdustinface replied:
Agreed, having it stored in each cache entry is weird. See f80e912b1d8cc17a576d8d8e0637dfe2e65e0ac7.

arvidn commented on tests/core/test_farmer_harvester_rpc.py:
```diff
@@ -169,37 +170,66 @@ async def have_plots():
 
+            # Reset cache and force updates cache every second to make sure the farmer gets the most recent data
+            update_interval_before = farmer_api.farmer.update_harvester_cache_interval
+            farmer_api.farmer.update_harvester_cache_interval = 2
```
I'm not convinced the increased complexity of sneaking in this parameter as a member here, to affect the behavior of this class later just for testing purposes carries its weight.

If it should be configurable, it should probably be a proper configuration option.

xdustinface replied:
I didn't put it into the config because it will be dropped soon anyway because i'm going to change the behavior so that the harvester reports the plots instead of the farmer requests them.

xdustinface commented:
Rebased on `main`, squashed https://github.com/Chia-Network/chia-blockchain/commit/f80e912b1d8cc17a576d8d8e0637dfe2e65e0ac7 and adjusted [this](https://github.com/Chia-Network/chia-blockchain/compare/f80e912b1d8cc17a576d8d8e0637dfe2e65e0ac7..bf1c6958733ab18f567c84aa215f28927b9e0292#diff-ebb107e316ce58dda857782dc62783bc29eb5c02e305e2f6c1ef44048f1b63feL175-R176).

arvidn commented on tests/core/test_farmer_harvester_rpc.py:
```diff
@@ -170,37 +171,66 @@ async def have_plots():
+                test_value("loaded_plots", refresh_result, expected_result)
+                test_value("removed_plots", refresh_result, expected_result)
+                test_value("processed_files", refresh_result, expected_result)
+                test_value("remaining_files", refresh_result, expected_result)
```
the left hand side here seems better. It's shorter and easier to read and it actually fails the unit test if the values disagree. As far as I can tell, you just set `expected_result_matched`. which is then overwritten by the next test. afaict, if the `loaded_plots` field is a mismatch, and the remaining ones are matches, the test will still pass...

xdustinface replied:
Why would it still pass? It gets set to `True` before the refresh gets triggered [here](https://github.com/Chia-Network/chia-blockchain/pull/8491/files#diff-ebb107e316ce58dda857782dc62783bc29eb5c02e305e2f6c1ef44048f1b63feR229) and asserted after the the refresh [here](https://github.com/Chia-Network/chia-blockchain/pull/8491/files#diff-ebb107e316ce58dda857782dc62783bc29eb5c02e305e2f6c1ef44048f1b63feR233). `test_value` does not overwrite it, its `test_case` which resets it.

xdustinface replied:
And even though the previous version was shorter, see this https://github.com/Chia-Network/chia-blockchain/pull/8491/commits/bf1c6958733ab18f567c84aa215f28927b9e0292 commit's message for the reason for this change .

arvidn replied:
I see, I misread the assignments to `=false`.

arvidn commented:
looks ok otherwise

arvidn commented:
I wonder if it would be better to turn this thread into a `ProcessPoolExecutor`. It won't be as good of a fit since you can then only submit jobs to it. But at least exceptions are correctly communicated back to the main thread.
